### DISCUSSION TITLE
Best approach for copying geometry from one iModel to another.

### DISCUSSION THREAD

**Discussion by spowell-216** [iTwin/itwinjs-core#114] [General] [answered]
What is the best approach for copying geometry from one iModel and inserting it into a new iModel?  
Currently using version 2.5.3,  I have the following method, but the "geom" from the existing part is returned as undefined. 

`private static async _insertPart(targetIMDb: IModelDb, sourceIMDb: IModelDb, spatialModelId: Id64String, partOID: string, spoolId: Id64String): Promise<Id64String> {
    try {
      // ECSQL query to get ECInstanceId for the source Part
      const oid = await this._queryByOID(sourceIMDb, partOID);

      // get that part from source
      const part = sourceIMDb.elements.getElement<GeometricElement3d>(oid);

      if (part.geom == undefined) console.log(`no geometry returned for this part: ${part.id}, Model: ${part.model}`);

      // recreate that part in Target
      const geometryPartProps: GeometricElement3dProps = {
        model: spatialModelId,
        code: Code.createEmpty(),
        classFullName: Class.Spool,
        category: Spool.getCategory(targetIMDb, Class.Spool).id,
        geom: part.geom,
        placement: part.placement,
        federationGuid: partOID,
        userLabel: part.userLabel,
        parent: new ElementOwnsChildElements(spoolId),
      };
      const partId = targetIMDb.elements.insertElement(geometryPartProps);
      return Promise.resolve(partId);
    } catch (err) {
      Logger.logError(IModelAgentConfig.loggingCategory, `error inserting part: ${err}`);
      return Promise.resolve("0");
    }
  }`

example of the debug log: 
no geometry returned for this part: 0x200000000f3, Model: 0x20000000002
no geometry returned for this part: 0x200000009e8, Model: 0x20000000002
no geometry returned for this part: 0x20000000fdd, Model: 0x20000000002
no geometry returned for this part: 0x20000000e1e, Model: 0x20000000002
no geometry returned for this part: 0x20000000c64, Model: 0x20000000002

but when i use iModelConsole and inspect for these ECInstanceIDs i see that there is a geometrystream defined
![image](https://user-images.githubusercontent.com/51784707/93909824-36bc6400-fcce-11ea-87af-bc8f3dc0ff8d.png)

**Comment by scsewall** [comment]
As a performance optimization, we don't load geometry unless it is specifically requested. If you look at the documentation for `getElement`, you will see that the `elementId` parameter can be of multiple types:

https://www.imodeljs.org/reference/imodeljs-backend/imodels/imodeldb.elements/getelement/?term=getelement

You probably want `ElementLoadProps` and to specify `wantGeometry: true`.

https://www.imodeljs.org/reference/imodeljs-common/entities/elementloadprops/

There are ways of *cloning* entire elements between iModels, but your code should work with the above changes. See this page if you want to learn more about cloning:

https://www.imodeljs.org/learning/backend/imodeltransformation/

Hope this helps.

**Comment by mgooding** [comment]
> You probably want ElementLoadProps and to specify wantGeometry: true.

You will also want to set `wantBRepData: true`.

**Comment by spowell-216** [comment, marked as answer]
The ElementLoadProps with wantGeometry and wantBRepData set to true resolved the undefined problem, but still not getting the geometry written to target iModel.  I parsed the geom to jsonProperties, for troubleshooting, and below is what is being captured.  

![image](https://user-images.githubusercontent.com/51784707/94022195-d4716b00-fd82-11ea-8a63-0aeb761c0955.png)

this looks to me to be a valid GeometryStreamProps, but I am guessing that all the referenced ECInstanceIds are source ids and I will have to find and recreate each element?  If i switch over to the IModelTransformer will this be handled automatically?

**Reply by scsewall** [reply]
You are correct. While it is possible to directly copy simple geometry primitives between iModels, most GeometryStreams also contain embedded references (local Ids) to DefinitionElements like materials and sub-categories for display appearance settings and to GeometryParts for reusable/shared geometry.

This is where the `IModelTransformer` API and its `IModelCloneContext` come in. Yes, IModelTransformer does handle the required source Id to target Id remapping, but there is configuration that needs to be in place before it is done *automatically*. Part of the prerequisite work includes importing cloned copies of the referenced DefinitionElements from the source into the target and setting up the remapping.

Could you please give a few more details on your scenario so that I can point you in the right direction?

**Reply by spowell-216** [reply]
Our scenario is that we have multiple iModels created from different bridges all with different schemas. Our goal is to have a centralized iModel with a normalized company defined schema.  

Another scenario we are investigating is based upon the sample-showcase 2D/3D integration.  For us we have 2D being generated in SmartPlant P&ID and 3D being generated in SmartPlant 3D.  Right now this is two separate bridges generating 2 sperate iModels and we are looking at how to integrate the 2 so we can reproduce the example of clicking on a 2D object on a P&ID and jump to it's location and highlight it in the 3D model.

Any help on these two scenarios would be gratefully appreciated.

**Reply by scsewall** [reply]
The first step would be to set up the desired structure and schema in the target iModel. The job after that is to set up the remapping rules to merge the different sources into the appropriate part in the structure of the target.

I would first point you to this unit test:

https://github.com/imodeljs/imodeljs/blob/ecb6c25d270eea736392961bd9ae4a0aea0bda66/core/backend/src/test/standalone/IModelTransformer.test.ts#L433

This test does not do a schema transform, but does demonstrate how to merge multiple source iModels into one target. Pay special attention to setting up a Subject in the target and setting the `targetScopeElementId` parameter when setting up the `IModelTransformer` instance.

This test demonstrates a schema transform:

https://github.com/imodeljs/imodeljs/blob/ecb6c25d270eea736392961bd9ae4a0aea0bda66/core/backend/src/test/standalone/IModelTransformer.test.ts#L42

 The source iModel has one schema and the target iModel has another and then the `onTransformElement` override maps between them. The `onTransformElement` method is here:

https://github.com/imodeljs/imodeljs/blob/ecb6c25d270eea736392961bd9ae4a0aea0bda66/core/backend/src/test/IModelTransformerUtils.ts#L1179

I am sure that you will have additional questions, but hopefully this will get you started.
